PurchasesCoreSwift/Identity/IdentityManager.swift

```diff
@@ -0,0 +1,160 @@
+        let currentAppUserIDLooksAnonymous = foundAnonymous
+
+        let isLegacyAnonymousAppUserID: Bool
+        if let cachedAppUserID = currentAppUserID {
+            isLegacyAnonymousAppUserID = cachedAppUserID == deviceCache.cachedLegacyAppUserID
+        } else {
+            isLegacyAnonymousAppUserID = false
+        }
+
+        return currentAppUserIDLooksAnonymous || isLegacyAnonymousAppUserID
+    }
+
+    private let deviceCache: DeviceCache
+    private let backend: Backend
+    private let purchaserInfoManager: PurchaserInfoManager
+
+    @objc public init(deviceCache: DeviceCache, backend: Backend, purchaserInfoManager: PurchaserInfoManager) {
+        self.deviceCache = deviceCache
+        self.backend = backend
+        self.purchaserInfoManager = purchaserInfoManager
+    }
+
+    @objc public func configure(appUserID maybeAppUserID: String?) {
+        let appUserID = maybeAppUserID
+            ?? deviceCache.cachedAppUserID
+            ?? deviceCache.cachedLegacyAppUserID
+            ?? generateRandomID()
+        Logger.user(String(format: Strings.identity.identifying_app_user_id, appUserID))
+
+        deviceCache.cache(appUserID: appUserID)
+        deviceCache.cleanupSubscriberAttributes()
+    }
+
+    @objc public func logIn(appUserID: String, completion: @escaping (PurchaserInfo?, Bool, Error?) -> Void) {
+        let newAppUserID = appUserID.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard let currentAppUserID = currentAppUserID,
+              !newAppUserID.isEmpty else {
+            let message = currentAppUserID == nil
+                ? Strings.identity.logging_in_with_initial_appuserid_nil
+                : Strings.identity.logging_in_with_nil_appuserid
+            Logger.error(message)
+            completion(nil, false, ErrorUtils.missingAppUserIDError())
+            return
+        }
+
+        guard newAppUserID != currentAppUserID else {
+            Logger.warn(Strings.identity.logging_in_with_same_appuserid)
+            purchaserInfoManager.purchaserInfo(appUserID: currentAppUserID) { maybePurchaserInfo, maybeError in
+                completion(maybePurchaserInfo, false, maybeError)
+            }
+            return
+        }
+
+        backend.logIn(currentAppUserID: currentAppUserID,
+                      newAppUserID: newAppUserID) { maybePurchaserInfo, created, maybeError in
+            if maybeError == nil,
+               let purchaserInfo = maybePurchaserInfo {
+                self.deviceCache.clearCaches(oldAppUserID: currentAppUserID, andSaveWithNewUserID: newAppUserID)
+                self.purchaserInfoManager.cache(purchaserInfo: purchaserInfo, appUserID: newAppUserID)
+            }
+
+            completion(maybePurchaserInfo, created, maybeError)
+        }
+    }
+
+    @objc public func logOut(completion: (Error?) -> Void) {
+        Logger.info(String(format: Strings.identity.logging_out_user, currentAppUserID ?? "<nil currentAppUserID>"))
```

Comment:
Again not migration-related, and I recognize this would be a change in logging behavior -- but as a user, I'd be confused seeing this error message (`Logging out user %@`) and then also seeing `logOutAnonymousUserError`, if no resetting/logging out is happening

Maybe we could change the string value to `Log out called for user %@` or something more clear?

Reply:
💯